### PR TITLE
Phase 1 

### PR DESCRIPTION
**Pull request by RJR-003**
Added code for button functionalities.
Wrote code for fetching data from Firebase and populating data in the table and dropdowns.
Wrote code for serch, sort, filter employees in typescript.

### CODE REVIEW

File: src/SortFun.ts [hunk 1]
```diff
@@ -0,0 +1,66 @@
+import { actualData, tableCreate } from "./script.ts";
+import { FilterArr } from "./filterAndSearchFun.ts";
+const tableBody = document.querySelector(".table-body")! as HTMLTableElement;
+
+let dirFlag = 1;
+const sortButton = document.querySelector(".sort-button")! as HTMLImageElement;
+// sort functionality
+//////////////////////////////////////
+const sortFun = () => {
+  console.log(FilterArr);
```
Comment (pratheeshrussell-qb) [inline src/SortFun.ts]: Remove unnecessary console logs before pushing

File: src/SortFun.ts [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+  let arrayToSort = actualData.employee;
+  if (FilterArr.length !== 0) arrayToSort = FilterArr;
+
+  let arrToRender = arrayToSort.sort(
+    (
+      a: {
```
Comment (pratheeshrussell-qb) [inline src/SortFun.ts]: extract custom types to a separate file and import the type here. Don't redefine types repeatedly

File: src/filterAndSearchFun.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+    }
+  });
+
+  const searchvalue = searchBar.value.toLowerCase();
+
+  let FilterArr = actualData.employee;
+
+  if (searchBar.value !== "") {
+    FilterArr = FilterArr.filter(
+      (elem: {
```
Comment (pratheeshrussell-qb) [inline src/filterAndSearchFun.ts]: extract custom types to different file

File: src/filterAndSearchFun.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+    );
+  }
+  tableBody.innerHTML = "";
+  tableCreate(FilterArr);
+};
+
+searchBar.addEventListener("input", filterTable);
+
+const changeSkillState = (skillId: string) => {
+  console.log(skillId);
```
Comment (pratheeshrussell-qb) [inline src/filterAndSearchFun.ts]: remove logs

File: src/script.ts [hunk 5]
```diff
@@ -0,0 +1,181 @@
+import { handleTableClick } from "./tableActionButton.ts";
+
+const api =
+  "https://hrm-app-39bd9-default-rtdb.asia-southeast1.firebasedatabase.app";
+
+export let actualData: {
+  employee: {
```
Comment (pratheeshrussell-qb) [inline src/script.ts]: extract type

File: src/script.ts [hunk 6]
```diff
@@ -0,0 +1,181 @@
+import { handleTableClick } from "./tableActionButton.ts";
+
+const api =
```
Comment (pratheeshrussell-qb) [inline src/script.ts]: it is a good idea to extract constants to a separate file. Although here there is only 1 constant there could be many in bigger apps

File: src/script.ts [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+        </td>
+    </tr>
+    `;
+  });
+};
+
+// fetching data from firebase and display it into the table
+///////////////////////////////////////////
+const fillentry = (obj: {
+  employee: {
```
Comment (pratheeshrussell-qb) [inline src/script.ts]: extract type definition

File: src/script.ts [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+};
+
+//fetching data whole data from firebase
+////////////////////////////////////////////////
+const fetchData = function (fillentry: Function): void {
+  fetch(api + "/.json")
+    .then((res) => res.json())
+    .then((data) => {
+      actualData = data;
+      console.log(data, "data");
```
Comment (pratheeshrussell-qb) [inline src/script.ts]: remove log

**Review by pratheeshrussell-qb — CHANGES_REQUESTED** [review]

**Comment by pratheeshrussell-qb** [comment]
@RJR-003  major points are

- extract the type definition to a separate file
- remove console logs added for debugging

I would also suggest adding a ts file named data handler service and add a singleton class that will do all data manipulations, like fetching, saving, updating. It will make the code neat and there wont be mixup of business logic and UI

File: src/constants.ts [hunk 9]
```diff
@@ -42,3 +44,4 @@ export const dataEntryClose = document.querySelector(
 export const dataEntryModal = document.querySelector(
```
Comment (pratheeshrussell-qb) [inline src/constants.ts]: why are query selectors added here? wont that be run in pages where they are not needed? maybe the selector string here but do the actual selection in the page where it is needed

**Review by pratheeshrussell-qb — CHANGES_REQUESTED** [review]

File: src/constants.ts [hunk 10]
```diff
@@ -0,0 +1,47 @@
+export const api =
```
Comment (pratheeshrussell-qb) [inline src/constants.ts]: A better and prefered way would be to export an Object with each key containing the const values (make sure to add as const declaration or you wont get type support).  Another way would be to export a class with all members marked as static. This way you can import a single item and use any const value from it

**Review by pratheeshrussell-qb — CHANGES_REQUESTED** [review]

File: src/type.ts [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+  skillID: number;
+  skill: string;
+};
+
+export type role = {
+  roleID: number;
+  role: string;
+};
+
+export type fullData = {
```
Comment (pratheeshrussell-qb) [inline src/type.ts:29]: fullData can be renamed as CompanyData or even as FirebaseData. "fulldata" is confusing

**Review by pratheeshrussell-qb — CHANGES_REQUESTED** [review]

File: src/deleteEmployee.ts [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+const confirmButton = document.querySelector(
+  ".confirm-button"
+)! as HTMLButtonElement;
+const delData = (index: number) => {
+  fetch(api + "/employee/" + index + ".json", {
+    method: "DELETE",
+  })
+    .then((res) => {
+      console.log(res, "successfully deleted!!!");
+      toast(false, "Succesfully deleted the employee");
```
Comment (pratheeshrussell-qb) [inline src/deleteEmployee.ts:24]: Extract all toast messages to constants. Add an object AppStrings and add it there

File: src/deleteEmployee.ts [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+import { api, skillList, overlay, dataDelModal } from "./constants";
+const confirmButton = document.querySelector(
+  ".confirm-button"
+)! as HTMLButtonElement;
+const delData = (index: number) => {
+  fetch(api + "/employee/" + index + ".json", {
+    method: "DELETE",
+  })
+    .then((res) => {
+      console.log(res, "successfully deleted!!!");
```
Comment (pratheeshrussell-qb) [inline src/deleteEmployee.ts:23]: Remove all unnecessary console logs. Except for getting error message codes in catch block you can remove other console logs

**Review by pratheeshrussell-qb — CHANGES_REQUESTED** [review]

File: src/deleteEmployee.ts (hunk 12 above)
Comment (pratheeshrussell-qb) [reply src/deleteEmployee.ts:24]: extract toast message in other files too

File: src/addUpdateEmployee.ts [hunk 14]
```diff
@@ -0,0 +1,326 @@
...
+  const target = e.target as HTMLElement;
+  if (target.classList.contains("edit-image-icon")) {
+    // skillNameArr = skillName;
+    changeSkillNameArr(skillName);
+  }
+};
+
+formSkill.onclick = (e: MouseEvent) => {
+  const target = e.target as HTMLElement;
+  // for chrome
```
Comment (pratheeshrussell-qb) [inline src/addUpdateEmployee.ts:304]: remove commented code - Adding commented descriptions as text is fine but don't add commented code

**Review by pratheeshrussell-qb — CHANGES_REQUESTED** [review]

File: src/constants.ts (hunk 9 above)
Comment (RJR-003) [reply src/constants.ts]: changed them and now given wherever necessary only

File: src/constants.ts (hunk 10 above)
Comment (RJR-003) [reply src/constants.ts]: made a constants object with key as the required constant

**Pull request closed without merging**